src/webutil/tableState.js

```diff
@@ -0,0 +1,273 @@
+      sortOrder: SortOrderTypes,
+      // A key representing the sort field. It is is used to toggle sort order
+      // and increase efficiency.
+      sortName: symbol,
+      // A function that returns the property of the element that should be used
+      // in the sort comparator.
+      sortFn: PluckFn<T>,
+    },
+  }
+): TableState<T> {
+  const optionDefaults = {
+    initialRowsPerPage: Infinity,
+    initialSort: {},
+    ...initialOptions,
+  };
+  const sortDefaults = {
+    sortOrder: SortOrders.ASC,
+    sortName: Symbol(),
+    sortFn: null,
+    ...optionDefaults.initialSort,
+  };
+  const [sortName, setSortName] = useState(sortDefaults.sortName);
+  const [sortOrder, setSortOrder] = useState<SortOrderTypes>(
+    sortDefaults.sortOrder
+  );
+  const [sortFn, setSortFn] = useState(() => sortDefaults.sortFn);
+  const [pageIndex, setPageIndex] = useState(0);
+  const [rowsPerPageUnsafe, setRowsPerPageInternal] = useState(
+    optionDefaults.initialRowsPerPage
+  );
+  const rowsPerPage = useMemo(() => {
+    if (rowsPerPageUnsafe !== Infinity) return rowsPerPageUnsafe;
+    else return data.length;
+  }, [rowsPerPageUnsafe, data]);
+  const [filterFns, setFilterFns] = useState<FilterFns<T>>({});
+
+  const sortedArray = useMemo(() => {
+    const array = data.slice();
+    if (sortFn) {
+      sortBy(array, sortFn);
+      if (sortOrder === SortOrders.DESC) array.reverse();
+    }
+    return array;
+  }, [data, sortFn, sortOrder]);
+
+  const filteredArray = useMemo(() => {
+    const reducer = (array, key) => {
+      const fn = filterFns[key];
+      if (fn) return array.filter(fn);
+      return array;
+    };
+    return Object.keys(filterFns).reduce(reducer, sortedArray);
+  }, [sortedArray, filterFns]);
+
+  const setRowsPerPage = useCallback(
+    (newRowsPerPage: number) => {
+      // Keep the user from losing their place
+      // when changing # of results per page
+      const intRowsPerPage = parseInt(newRowsPerPage, 10);
+      const currentFirstVisibleRow = pageIndex * rowsPerPage;
+      const newPage = Math.floor(currentFirstVisibleRow / intRowsPerPage);
+
+      setRowsPerPageInternal(intRowsPerPage);
+      setPageIndex(newPage);
+    },
+    [pageIndex, rowsPerPage, data]
+  );
+
+  const setSortFnCallback = useCallback(
+    (newSortName: symbol, newSortFn: PluckFn<T>) => {
+      if (newSortName === sortName) {
+        setSortOrder((currentSortOrder) =>
+          currentSortOrder === SortOrders.DESC
+            ? SortOrders.ASC
+            : SortOrders.DESC
+        );
+      } else {
+        setSortOrder(DEFAULT_SORT);
+        setSortFn(() => newSortFn);
+      }
+      setSortName(newSortName);
+    },
+    [sortName]
+  );
+
+  const createOrUpdateFilterFn = useCallback(
+    (key: string, newFilterFn: null | ((T) => boolean)) => {
+      setPageIndex(0);
+      setFilterFns((fns): FilterFns<T> => {
+        return {
+          ...fns,
+          [key]: newFilterFn,
+        };
```

Comment:
unfortunately, I was not able to get symbols to work for the filterFns. I had to use `// $FlowFixMe` since symbols are not supported as keys by flow, and even then, it didn't work. I'd like to continue using strings as a good-enough solution, with hopes to maybe figure out why symbols don't work in the future.